# EES1133 Lab 1: Simple Greenhouse Model

Adapted from notes by:
[Brian E. J. Rose](http://www.atmos.albany.edu/facstaff/brose/index.html), University at Albany

In this Lab, we will mainly be using python as a calculator. I have left a few things blank for you to fill in. So, follow along and fill in the variables as you go. When you are done, complete the lab quiz on Quercus.

## The observed global energy budget

Let's take a look at the observational global energy budget (we will explore this figure more next lecture):

![Observed global energy flows from Trenberth and Fasullo (2012)](https://github.com/kls2177/ccia_files/blob/master/GlobalEnergyBudget.png?raw=true)

### Recap of our simple greenhouse model

In class we introduced a very simple model for the **OLR** or Outgoing Longwave Radiation to space:

$$ \text{OLR} = \epsilon \sigma T_s^4 $$

where $\epsilon$ is the **emissivity** of the atmosphere, a number less than 1 that represents the greenhouse effect of Earth's atmosphere.

We also tuned this model to the observations by choosing $ \epsilon \approx 0.61$.

More precisely:

In [1]:
# define variables
OLR_observed = 238.5  # in W/m2 (see figure above)
sigma = 5.67E-8  # Stefan-Boltzman constant
Ts_observed = 288.  # global average surface temperature

In [2]:
# solve for tuned value of emissivity
epsilon = OLR_observed / sigma / Ts_observed**4  
print(epsilon)

0.6114139923687016


Let's now deal with the shortwave (solar) side of the energy budget.

### Absorbed Shortwave Radiation (ASR) and Planetary Albedo

Let's define a few terms.

#### Global mean insolation

From the observations (see figure above), the area-averaged incoming solar radiation or **insolation** is 341.3 W m$^{-2}$ (i.e., $So$/4 or 1365/4 W m$^{-2}$).

Let's denote this quantity by $Q$. Fill in the cell below.

In [6]:
Q =   # the insolation, i.e. So/4

#### Planetary albedo

Some of the incoming radiation is not absorbed at all but simply reflected back to space. Let's call this quantity $F_{reflected}$

From observations we can fill in the cell below (see figure above):

In [5]:
F_reflected = #  reflected shortwave flux in W/m2


The **planetary albedo** is the fraction of $Q$ that is reflected.

We will denote the planetary albedo by $\alpha$.

From the observations we can fill in the cell below:

In [16]:
alpha = 
print(alpha)

That is, about 30% of the incoming radiation is reflected back to space.

#### Absorbed Shortwave Radiation

The **Absorbed Shortwave Radiation** or ASR is the part of the incoming sunlight that is *not* reflected back to space, i.e. that part that is absorbed somewhere within the Earth system (at the surface or in the atmosphere).

Mathematically, we write

$$ \text{ASR} = Q - F_{reflected} = (1-\alpha) Q $$

From the observations we can fill in the cell below:

In [17]:
# compute ASR
ASR_observed = 
print(ASR_observed)

This number is *just slightly greater* than the observed OLR of 238.5 W m$^{-2}$.

____________
<a id='section3'></a>
## I. Equilibrium temperature
____________

*This is one of the central concepts in climate modeling.*

The Earth system is in **energy balance** when energy in = energy out, i.e. when

$$ \text{ASR} = \text{OLR} $$

We want to know:

- What surface temperature do we need to have this balance?
- By how much would the temperature change in response to other changes in Earth system?
    - Changes in greenhouse gases
    - Changes in cloudiness
    - etc.




With our simple **greenhouse model**, we can get an **exact solution** for the equilibrium temperature.

First, write down our statement of energy balance:

$$ (1-\alpha) Q = \epsilon \sigma T_s^4 $$

Rearrange to solve for $T_s$:

$$ T_s^4 = \frac{(1-\alpha) Q}{\epsilon \sigma} $$

and take the fourth root, denoting our **equilibrium temperature** as $T_{eq}$:

$$ T_{eq} = \left( \frac{(1-\alpha) Q}{\epsilon \sigma} \right)^\frac{1}{4} $$

Let's define a resuable function to compute $T_{eq}$:

In [18]:
# define a reusable function based on the above equation!
def equilibrium_temperature(alpha,Q,epsilon): 
    return ((1-alpha)*Q/(epsilon*sigma))**(1/4)

Complete the cell below by plugging the observed values for alpha, Q and epsilon:

In [3]:
# use your function to compute Teq

Teq_observed = equilibrium_temperature()
print(Teq_observed)

### Reflection Question:
This equilibrium temperature is *just slightly warmer* than 288 K.  Why?

### A climate change scenario

Suppose that, due to global warming (changes in atmospheric composition and subsequent changes in cloudiness):

- The longwave emissivity decreases to $\epsilon = 0.57$ 
- The planetary albedo increases to $\alpha = 0.32$

What is the new equilibrium temperature?

For this very simple model, we can work out the answer exactly:

In [16]:
Teq_new = equilibrium_temperature(0.32,Q,0.57)

#  an example of formatted print output, limiting to two or one decimal places
print('The new equilibrium temperature is {:.2f} K.'.format(Teq_new))
print('The equilibrium temperature increased by about {:.1f} K.'.format(Teq_new-Teq_observed))

The new equilibrium temperature is 291.04 K.
The equilibrium temperature increased by about 2.9 K.


Most climate models are more complicated mathematically, and solving directly for the equilibrium temperature will not be possible! 

Instead, we will be able to use the model to calculate the terms in the energy budget (ASR and OLR).

### Python exercise (optional)

- Write Python functions to calculate ASR and OLR for *arbitrary parameter values*.
- Verify the following:
    - With the new parameter values but the old temperature T = 288 K, is ASR greater or lesser than OLR? 
    - Is the Earth gaining or losing energy?
    - How does your answer change if T = 295 K (or any other temperature greater than 291 K)?

In [ ]:
# define a reusable function for ASR and OLR, separately! The input for the ASR function
# should be alpha and the inputs for the OLR function should be epsilon and Ts

def ASR(alpha):
    #fill this in
    return ASR

In [ ]:
def OLR(epsilon,Ts):
    #fill this in
    return OLR

In [19]:
print(ASR(0.32),OLR(0.57,295))

____________
<a id='section3'></a>
## II. A time-dependent Energy Balance Model
____________

The above exercise shows us that if some properties of the climate system change in such a way that the **equilibrium temperature goes up**, then the Earth system *receives more energy from the sun than it is losing to space*. The system is **no longer in energy balance**.

The temperature must then increase to get back into balance. The increase will not happen all at once! It will take time for energy to accumulate in the climate system. We want to model this **time-dependent adjustment** of the system.

In fact almost all climate models are **time-dependent**, meaning the model calculates **time derivatives** (rates of change) of climate variables.

### An energy balance equation

We will write the **total energy budget** of the Earth system as

$$ \frac{dE}{dt} = \text{ASR} - \text{OLR} $$

**This equation is the starting point for EVERY CLIMATE MODEL.**

But so far, we don’t actually have a MODEL. We just have a statement of a budget. To use this budget to make a model, we need to relate terms in the budget to state variables of the atmosphere-ocean system.

For now, the state variable we are most interested in is **temperature** – because it is directly connected to the physics of each term above.

Suppose that we can write $E$ as:

$$ E = C T_s $$

Our budget then becomes:

$$ C \frac{dT_s}{dt} = \text{ASR} - \text{OLR} $$

where

- $C$ is the **heat capacity** of Earth system, in units of J m$^{-2}$ K$^{-1}$.
- $\frac{dT_s}{dt}$ is the rate of change of global average surface temperature.

By adopting this equation, we are assuming that the energy content of the Earth system (atmosphere, ocean, ice, etc.) is *proportional to surface temperature*.

Important things to think about:

- Why is this a sensible assumption?
- What determines the heat capacity $C$?
- What are some limitations of this assumption?


For our purposes here we are going to use a value of $C$ equivalent to heating 100 meters of water:

In [20]:
c_w = 4E3             #  Specific heat of water in J/kg/K
rho_w = 1E3           #  Density of water in kg/m3
H = 100.              #  Depth of water in m
C = c_w * rho_w * H   #  Heat capacity of the model 
print('The effective heat capacity is {:.1e} J/m2/K'.format(C))

The effective heat capacity is 4.0e+08 J/m2/K


## Representing time derivatives on a computer

Recall that the derivative is the **instantaneous rate of change**. It is defined as 

$$ \frac{dT}{dt} = \lim_{\Delta t\rightarrow 0}⁡ \frac{\Delta T}{\Delta t}$$

- **On the computer there is no such thing as an instantaneous change.** 
- We are always dealing with *discrete quantities*.
- So we approximate the derivative with $\Delta T/ \Delta t$. 
- So long as we take the time interval $\Delta t$ "small enough", the approximation is valid and useful.
- (The meaning of "small enough" varies widely in practice. Let's not talk about it now)

So we write our model as:

$$ C  \frac{\Delta T}{\Delta t} \approx \text{ASR} - \text{OLR}$$

where $\Delta T$ is the **change in temperature predicted by our model** over a short time interval $\Delta t$.

We can now use this to **make a prediction**: 

Given a current temperature $T_1$ at time $t_1$, what is the temperature $T_2$ at a future time $t_2$?

We can write

$$ \Delta T = T_2-T_1 $$
$$ \Delta t = t_2-t_1 $$

and so our model says

$$ C  \frac{T_2-T_1}{\Delta t} = \text{ASR} - \text{OLR} $$

Which we can rearrange to **solve for the future temperature**:

$$ T_2 = T_1 + \frac{\Delta t}{C} \left( \text{ASR} - \text{OLR}(T_1) \right)  $$

We now have a formula with which to make our prediction!

Notice that we have written the OLR as a *function of temperature*. We will use the current temperature $T_1$ to compute the OLR, and use that OLR to determine the future temperature.

## Numerical solution of the Energy Balance Model

The quantity $\Delta t$ is called a **timestep**. It is the smallest time interval represented in our model.

Here we're going to use a timestep of 1 year. What units should $\Delta t$ be in?

In [21]:
# Fill in and run this cell
dt =    # one year expressed in seconds

In [2]:
# Let's try a single timestep, assuming we have working functions for ASR and OLR. Use alpha = 0.32 and epsilon = 0.57.
T1 = 288
T2 = T1 + dt / C * ( ASR(0.32) - OLR(0.57,T1) )
print(T2)

What happened? Why?

Try another timestep

In [20]:
T1 = 
T2 = 
print(T2)

Warmed up again, but by a smaller amount.

But this is tedious typing. Time to **define a function** with temperature as the input to make things easier and more reliable:

In [12]:
def step_forward(T):

Try it out with an arbitrary temperature:

In [13]:
step_forward()

Notice that our function calls other functions and variables we have already defined.

#### Python fact: Functions can access variables and other functions defined outside of the function. 

This is both very useful and occasionally confusing.

Now let's really harness the power of the computer by **making a loop** (and storing values in arrays):

In [1]:
import numpy as np

numsteps = 100
Tsteps = np.zeros(numsteps+1) # initialize array
Years = np.zeros(numsteps+1)  # initialize arrya
Tsteps[0] = 288. # initialize temperature

# loop over number of years (default starts at zero; ends at numsteps (100))
for n in range(numsteps): 
    Years[n+1] = n+1 # populate the Years array - add one each time to count the number of years
    Tsteps[n+1] = step_forward( Tsteps[n] ) # populate the Tsteps array - use our step_forward function each time
print(np.round(Tsteps,3)) # print temperature and round to nearest thousandth

What did we just do?

- Created an array of zeros
- set the initial temperature to 288 K
- repeated our time step 100 times. 
- Stored the results of each time step into the array.

#### Python fact: the `for` statement executes a statement (or series of statements) a specified number of times (a loop!)

#### Python fact: Use square bracket [ ] to refer to elements of an array or list. Use round parentheses ( ) for function arguments. 

### Plotting the result

Now let's draw a picture of our result!

In [16]:
#  import the plotting package
import matplotlib.pyplot as plt

Run the cell below to make a plot. We are using the `plt.plot()` function that plots a line plot.

In [17]:
plt.plot(Years, Tsteps )
plt.xlabel('Years')
plt.ylabel('Global mean temperature (K)');

Note how the temperature *adjusts smoothly toward the equilibrium temperature*, that is, the temperature at which
ASR = OLR.

**If the planetary energy budget is out of balance, the temperature must change so that the OLR gets closer to the ASR!**

The adjustment is actually an *exponential decay* process: The rate of adjustment slows as the temperature approaches equilibrium. 

The temperature gets very very close to equilibrium but never reaches it exactly.

#### Python fact: We can easily make simple graphs with the function plot(x,y), where x and y are arrays of the same size. But we must import it first. 

This is actually not native Python, but uses a special graphics library called `matplotlib`. 

Just about all of our notebooks will start with this:
```
import numpy as np
import matplotlib.pyplot as plt
```

____________
<a id='section3'></a>
## Summary
____________

- We looked at the flows of energy in and out of the Earth system. 
- These are determined by radiation at the top of the Earth's atmosphere.
- Any imbalance between shortwave absorption (ASR) and longwave emission (OLR) drives a change in temperature
- We built a climate model!
- This **Zero-Dimensional Energy Balance Model** solves for the global, annual mean surface temperature $T_s$
- Two key assumptions:
    - Energy content of the Earth system varies proportionally to $T_s$
    - The OLR increases as $\epsilon \sigma T_s^4$ (our simple greenhouse model)
- The model has an **equilibrium temperature** $T_{eq}$ at which ASR = OLR
- If $T_s < T_{eq}$, the model will warm up.
- We can represent the **continous warming process** on the computer using **discrete timesteps**.
- We can plot the result.

____________
<a id='section3'></a>
## Questions
____________

1. a. How long does it take our simple model to reach equilibrium ($\epsilon$=0.57, $\alpha$=0.32 and the initial temperature is 288K). Hint: add the variable "Years" to your print statment in your loop above? Assume that equilibrium is reached when the temperature stabilizes to the nearest thousandth.
<p/>
    b. What is the final equilibrium temperature?
<p/>
</p>

2. a. What happens if you change the initial temperature from 288K to 300K?
<p/>
    b. How long does it take to reach equilibrium now? Assume that equilibrium is reached when the temperature stabilizes to the nearest thousandth. What is the final equilibrium temperature?
<p/>
</p>

3. a. Starting from an initial temperature of 288K, what happens if you change the specific heat capacity, change H from 100m to 50m. Why?
<p/>
    b. Now suppose H is 500m - what happens? Why?
<p/>
</p>
    
4. a. Again, starting from an initial temperature of 288K, what happens if you change the time step from 1 year to 10 years? Why?
<p/>
</p>